### PR TITLE
Fix (de)compression

### LINKED ISSUE #294: Broken dxtbx.dlsnxs2cbf output
```
$ dxtbx.dlsnxs2cbf $(dials.data get -q thaumatin_eiger_screen)/Therm_6_1_master.h5 out_%04d.cbf
$ dials.image_viewer out_*.cbf
```
Image 1:
Screenshot 2021-01-29 at 16 22 34
Image 2:
Screenshot 2021-01-29 at 16 22 44
Image 3:
Screenshot 2021-01-29 at 16 22 58

### PR DESCRIPTION
**Pull request by graeme-winter**
Fix to errors in the byte-offset (de)compression:

- 0xf777 -> 0x7fff as one greater than -0x8000
- treatment of the fall through / last case in compression for similarity with decompression layout
- more thorough testing in particular tests with negative delta of magnitude greater than 0x8000
- test added which verifies that the HDF5 and CBF representation of the data from dlsnxs2cbf are the same

Adds tests as well

And fixes #294

Previously the code was working so pretty bad it got this far

### CODE REVIEW

File: tests/test_compression.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+def test_compress_decompress():
+    x, y = 10, 10
+
+    original = [1 for n in range(x * y)]
+
+    original[10] = 44369
+    original[11] = 214
+
+    data = flex.int(original)
```
Comment (rjgildea) [inline tests/test_compression.py:14]:
```suggestion
    data = flex.int(x * y, 1)
    data[10] = 44369
    data[11] = 214
```
You can construct a `flex.int` full of `1`s directly.

File: tests/test_compression.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    compressed = compress(data)
+    uncompressed = uncompress(compressed, x, y)
+
+    for d, u in zip(data, uncompressed):
+        assert d == u
```
Comment (rjgildea) [inline tests/test_compression.py:19]: You can use `pytest.approx` to compare values in one line:
```suggestion
    import pytest
    data = pytest.approx(uncompressed)
```

File: newsfragments/314.bugfix [hunk 3]
```diff
@@ -0,0 +1 @@
+CBF (de)compression: fix bugs in image compression and decompression
```
Comment (rjgildea) [inline newsfragments/314.bugfix:1]: Could you expand on under what circumstances this bug is applicable? As far as I am aware it was only (noticed) in `dlsnxs2cbf` - could it affect reading "normal" cbfs?

**Review by rjgildea — APPROVED** [review]
I haven't tested this so I assume the code is now behaving as expected. I think the newsfragment could do with being more explicit as to what the problem was/what was affected. Could you also please include a more verbose description of what the actual bug(s) was in the commit message when squash/merging? This would probably be more helpful than trying to infer via the diffs.

**Comment by graeme-winter** [comment]
Proper final commit message:

Fix to errors in the byte-offset (de)compression:
 - 0xf777 -> 0x7fff as one greater than -0x8000
 - treatment of the fall through / last case in decompression for similarity with compression layout
 - more thorough testing in particular tests with negative delta of magnitude greater than 0x8000
 - test added which verifies that the HDF5 and CBF representation of the data from `dlsnxs2cbf` are the same

File: newsfragments/314.bugfix (hunk 3 above)
Comment (graeme-winter) [reply newsfragments/314.bugfix]: More news added. I don't know yet if this could affect reading of "normal" CBF images, I will re-review in a moment.

Comment (graeme-winter) [reply newsfragments/314.bugfix]: Re-reviewed
